File: src/main/java/com/example/usercrud/business/config/aspect/LoggingAspect.java
package com.example.usercrud.business.config.aspect;

import lombok.extern.slf4j.Slf4j;
import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.annotation.AfterReturning;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.aspectj.lang.annotation.Pointcut;
import org.springframework.stereotype.Component;
import java.util.Arrays;
import java.util.concurrent.atomic.AtomicReference;

@Component
@Aspect
@Slf4j
public class LoggingAspect {
    @Pointcut("@annotation(com.example.usercrud.business.entity.annotations.Loggable)")
    public void methodLevel(){

    }

    @Pointcut("within(@com.example.usercrud.business.entity.annotations.Loggable *)")
    public void beanAnnotatedWithLoggable(){

    }

    @Pointcut("execution(public * *(..))")
    public void publicMethod() {

    }

    @Pointcut("publicMethod() && beanAnnotatedWithLoggable()")
    public void publicMethodInsideAClassAnnotatedWithLoggable(){

    }

    @Before("methodLevel() || publicMethodInsideAClassAnnotatedWithLoggable()")
    public void beforeMethodCall(JoinPoint joinPoint) {
        String className = joinPoint.getTarget().getClass().getName();
        String methodName = joinPoint.getSignature().getName();
        AtomicReference<String> args = new AtomicReference<>("");
        Arrays.stream(joinPoint.getArgs()).forEach(arg -> args.updateAndGet(v -> v + arg));
        log.info("Class name: {}, method name: {}, arguments: {}", className, methodName, args.get());
    }

    @AfterReturning(pointcut = "methodLevel() || publicMethodInsideAClassAnnotatedWithLoggable()", returning = "returnValue")
    public void afterMethodCall(JoinPoint joinPoint, Object returnValue) {
        log.info("Returning value: {}", returnValue);
    }

}

File: src/main/kotlin/cloud/folium/usercrud/business/aspect/LoggingAspect.kt
package cloud.folium.usercrud.business.aspect

import org.aspectj.lang.JoinPoint
import org.aspectj.lang.annotation.AfterReturning
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Before
import org.aspectj.lang.annotation.Pointcut
import org.springframework.stereotype.Component
import java.util.*
import java.util.concurrent.atomic.AtomicReference

@Component
@Aspect
class LoggingAspect {
    @Pointcut("@annotation(cloud.folium.usercrud.business.entity.annotations.Loggable)")
    fun methodLevel() {
    }

    @Pointcut("within(@cloud.folium.usercrud.business.entity.annotations.Loggable *)")
    fun beanAnnotatedWithLoggable() {
    }

    @Pointcut("execution(public * *(..))")
    fun publicMethod() {
    }

    @Pointcut("publicMethod() && beanAnnotatedWithLoggable()")
    fun publicMethodInsideAClassAnnotatedWithLoggable() {
    }

    @Before("methodLevel() || publicMethodInsideAClassAnnotatedWithLoggable()")
    fun beforeMethodCall(joinPoint: JoinPoint) {
        val className = joinPoint.target.javaClass.name
        val methodName = joinPoint.signature.name
        val args = AtomicReference("")
        Arrays.stream(joinPoint.args).forEach { arg: Any -> args.updateAndGet { v: String -> v + arg } }
        //log.info("Class name: {}, method name: {}, arguments: {}", className, methodName, args.get())
    }

    @AfterReturning(
        pointcut = "methodLevel() || publicMethodInsideAClassAnnotatedWithLoggable()",
        returning = "returnValue"
    )
    fun afterMethodCall(joinPoint: JoinPoint?, returnValue: Any?) {
        //LoggingAspect.log.info("Returning value: {}", returnValue)
    }
}